contracts/nois-airdrop/src/contract.rs

```diff
@@ -0,0 +1,790 @@
+        exists.manager = tmp_manager;
+        Ok(exists)
+    })?;
+
+    Ok(Response::new().add_attribute("action", "update_config"))
+}
+
+// This function will call the proxy and ask for the randomness round
+pub fn execute_rand_drop(
+    deps: DepsMut,
+    _env: Env,
+    info: MessageInfo,
+    random_beacon_after: Timestamp,
+) -> Result<Response, ContractError> {
+    // Make sure the manager is executing this function
+    let config = CONFIG.load(deps.storage)?;
+    let manager = config.manager.ok_or(ContractError::Unauthorized {})?;
+    if info.sender != manager {
+        return Err(ContractError::Unauthorized {});
+    }
+
+    // For full transparency make sure the merkle root has been registered before beacon request
+    if MERKLE_ROOT.may_load(deps.storage)?.is_none() {
+        return Err(ContractError::MerkleRootAbsent);
+    }
+
+    let RandomnessParams {
+        nois_proxy,
+        nois_randomness,
+        requested,
+    } = NOIS_RANDOMNESS.load(deps.storage).unwrap();
+    // Prevents requesting randomness twice.
+    if requested {
+        return Err(ContractError::ImmutableRandomness);
+    }
+    NOIS_RANDOMNESS.save(
+        deps.storage,
+        &RandomnessParams {
+            nois_proxy: nois_proxy.clone(),
+            nois_randomness,
+            requested: true,
+        },
+    )?;
+
+    let response = Response::new().add_message(WasmMsg::Execute {
+        contract_addr: nois_proxy.into(),
+        // GetRandomnessAfter requests the randomness from the proxy after a specific timestamp
+        // The job id is needed to know what randomness we are referring to upon reception in the callback.
+        // In this example we only need 1 random number so this can be hardcoded to "airdrop"
+        msg: to_binary(&ProxyExecuteMsg::GetRandomnessAfter {
+            after: random_beacon_after,
+            job_id: "airdrop".to_string(),
+        })?,
+        // We pay here the proxy contract with whatever the manager sends. The manager needs to check in advance the proxy prices.
+        funds: info.funds, // Just pass on all funds we got
+    });
+    Ok(response)
+}
+
+pub fn execute_receive(
+    deps: DepsMut,
+    _env: Env,
+    info: MessageInfo,
+    callback: NoisCallback,
+) -> Result<Response, ContractError> {
+    let RandomnessParams {
+        nois_proxy,
+        nois_randomness: _,
+        requested,
+    } = NOIS_RANDOMNESS.load(deps.storage).unwrap();
+
+    // callback should only be allowed to be called by the proxy contract
+    // otherwise anyone can cut the randomness workflow and cheat the randomness by sending the randomness directly to this contract
+    ensure_eq!(info.sender, nois_proxy, ContractError::UnauthorizedReceive);
+    let randomness: [u8; 32] = callback
+        .randomness
+        .to_array()
+        .map_err(|_| ContractError::InvalidRandomness)?;
+    // Make sure the randomness does not exist yet
```

Comment:
You can do this in the beginning of the function where you load NOIS_RANDOMNESS already. No need to load again.